Removed code deprecated in 0.22 [transpiler and providers]

santhosh-gnse opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ *] I have added the tests to cover my changes.
- [* ] I have updated the documentation accordingly.
- [* ] I have read the CONTRIBUTING document.
-->

### Summary

	* Removed the argument 'qubit_channel_mapping' in the class 'qiskit.transpiler.passes.calibration.rzx_builder.RZXCalibrationBuilder'

	* Replaced the argument 'qobj[Qobj]' in 'qiskit.providers.aer.QasmSimulator.run()' with 'qcirc[QuantumCircuit]'

	* Altered the test files accordingly

	* Added a release note



### Details and comments

This PR addresses issue #10813. It involves the removal of the `qubit_channel_mapping` argument from the RZXCalibrationBuilder class and provides a migration example in the release notes. Additionally, this also replaces the qobj argument [Qobj object], has been replaced with qcirc [QuantumCircuit object]. Tests that conflicted with these changes have been modified to accommodate this adjustment.

The test file `test_basicaer_qobj_headers.py` has been removed, which previously checked for the header component passed into the qobj object and whether it could be retrieved through the result object. Since we are removing the qobj itself, this test can be disregarded, hence it has been deleted, and corresponding migration instructions have been added to the release notes.

This PR has been requested to tagged as [`Changelog: Removal`](https://github.com/Qiskit/qiskit/labels/Changelog%3A%20Removal) to reflect the removal deprecated components.

santhosh-gnse committed:
Removed code deprecated in 0.22 [transpiler and providers] (#10813)

	* Removed the argument 'qubit_channel_mapping' in the class 'qiskit.transpiler.passes.calibration.rzx_builder.RZXCalibrationBuilder'

	* Replaced the argument 'qobj[Qobj]' in 'qiskit.providers.aer.QasmSimulator.run()' with 'qcirc[QuantumCircuit]'

	* Altered the test files accordingly

	* Added a release note

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=10872) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=10872) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit?pullRequest=10872) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit?pullRequest=10872) it.</sub>

ElePT commented on releasenotes/notes/remove-deprecated-code-in-0.22.0-139fa09ee0cc6df9.yaml:
```diff
@@ -0,0 +1,36 @@
+      # Create a target object for the Aer simulator
+      target = providers.aer.QasmSimulator().target()
+      # Create an RZXCalibrationBuilder object using the target object
+      builder = RZXCalibrationBuilder(target=target)
```
I believe that not specifying these lines as code blocks is confusing Sphinx and breaking the lint and docs build (see the [docs workflow](https://dev.azure.com/qiskit-ci/qiskit/_build/results?buildId=49924&view=logs&j=abfb3e93-d1ff-5b1b-cb4c-74d6dcf536f9&t=60edc4b3-4259-59c8-ae13-dc31febe2d3d&l=254)). While you can add code blocks with the following syntax:
```
    Here is some reno text... For example:

    .. code-block:: python
    
      # This is a comment
      # Respecting the indentation of the code block is important
      x = 1
```
I think that this example may actually confuse users, as it looks like `target` is a direct replacement for `qubit_channel_mapping`, when this not the case. Given that this was a deprecation without replacement, I'd say that we don't need a migration example here. I would directly remove it and keep it simple.

santhosh-gnse replied:
Thank you so much for the feedback. Let me update the .yaml file and push the commit

ElePT commented on qiskit/providers/basicaer/qasm_simulator.py:
```diff
@@ -387,18 +385,11 @@ def _validate_measure_sampling(self, experiment):
         Args:
-            qobj (Qobj): payload of the experiment
+            qcirc (Union[QuantumCircuit, List[QuantumCircuit]]): circuit of the experiment
             backend_options (dict): backend options
```
From what I have understood, this deprecation was introduced when the interface of `BasicAer` backends (for example, `QasmSimulatorPy`) was updated to follow the interface of `BackendV1`. If you look at the `BackendV1` class [(link)](https://github.com/Qiskit/qiskit/blob/main/qiskit/providers/backend.py#L213), the `run` method has a `run_input` field where you input the circuit, so it would probably make sense to use the name `run_input` instead of `qcirc` here.

santhosh-gnse replied:
Thank you so much for pointing it out. I will make the necessary edits and push the commit for review.

ElePT commented on releasenotes/notes/remove-deprecated-code-in-0.22.0-139fa09ee0cc6df9.yaml:
```diff
@@ -0,0 +1,36 @@
+      simulator = QasmSimulator()
+      transpiled_circuit = transpile(qc, simulator)
+      # Run the simulation
+      job = simulator.run(transpiled_circuit, shots=1024)
```
I would also keep this second section very simple, the deprecation was already explained in the original PR (see reno [here](https://github.com/Qiskit/qiskit/pull/5128/files#diff-1c7f90e0c4c55483edffcd15b5f3ef23dbfeb8ed05b5267be17d44369425bc9e)). So here, I would get rid of the "old" section and just mention that the deprecated `qobj` input has been removed and an example with the new use (in a code block as explained above).

santhosh-gnse replied:
Definitely will do and thank you so much for your continuous feedback. Being new to the OS contributions, all your suggestions were invaluable for me.

ElePT reviewed (commented):
Thank you very much for your contribution. I took a look at the PR and left a few comments, let me know if anything is unclear :)

santhosh-gnse committed:
Merge branch 'main' of https://github.com/Matrixmang0/qiskit into remove-deprecate

santhosh-gnse committed:
Update release note and change input argument variable name

santhosh-gnse committed:
Merge branch 'main' into remove-deprecate

ElePT commented on qiskit/providers/basicaer/qasm_simulator.py:
```diff
@@ -387,18 +385,21 @@ def _validate_measure_sampling(self, experiment):
+                sure to update the docstrings of subclasses implementing this
+                method to document that. New provider implementations should not
+                do this though as :mod:`qiskit.qobj` will be deprecated and
+                removed along with the legacy providers interface.
```
This part of the docstring doesn't look necessary here.

santhosh-gnse replied:
Thank you once more for your assistance. I too felt the same, let me retain the original wording in the string. Your help is greatly appreciated.

ElePT reviewed (commented):
Thanks a lot! I have added a small comment, but the changes look good to me. Because I am not that familiar with the original deprecation. I'd rather have a second reviewer check that this is the intended change (@1ucian0), especially with the `qobj` -> `run_input` rename.

santhosh-gnse committed:
Merge branch 'Qiskit:main' into remove-deprecate

santhosh-gnse committed:
Merge branch 'main' of https://github.com/Matrixmang0/qiskit into remove-deprecate

santhosh-gnse committed:
Merge branch 'remove-deprecate' of https://github.com/Matrixmang0/qiskit into remove-deprecate

santhosh-gnse committed:
Edited the docstring for the arg run_input

1ucian0 commented on qiskit/providers/basicaer/qasm_simulator.py:
```diff
@@ -411,29 +402,26 @@ def run(self, qobj, **backend_options):
             initial statevector for the simulator to be used instead of the all
             zero state. This size of this vector must be correct for the number
-            of qubits in all experiments in the qobj.
+            of qubits in run_input.
```
```suggestion
            of qubits in ``run_input`` parameter.
```

1ucian0 commented on releasenotes/notes/remove-deprecated-code-in-0.22.0-139fa09ee0cc6df9.yaml:
```diff
@@ -0,0 +1,30 @@
+upgrade:
+  - |
+    Removed the argument `qubit_channel_mapping` in :class:`qiskit.transpiler.passes.calibration.rzx_builder.RZXCalibrationBuilder`
+
```
```suggestion
    Removed the argument `qubit_channel_mapping` in :class:`qiskit.transpiler.passes.calibration.rzx_builder.RZXCalibrationBuilder`,
    which was deprecated in Qiskit 0.39 (released on Oct 2022, with qiskit-terra 0.22)
```

1ucian0 commented on releasenotes/notes/remove-deprecated-code-in-0.22.0-139fa09ee0cc6df9.yaml:
```diff
@@ -0,0 +1,30 @@
+    Removed the argument `qubit_channel_mapping` in :class:`qiskit.transpiler.passes.calibration.rzx_builder.RZXCalibrationBuilder`
+
+  - |
+    Replaced the argument `qobj[Qobj]` in :meth:`qiskit.providers.aer.QasmSimulator.run()` with `run_input[QuantumCircuit or Schedule or list]`
```
```suggestion
    Replaced the argument ``qobj[Qobj]`` in :meth:`qiskit.providers.aer.QasmSimulator.run()` with ``run_input[QuantumCircuit or Schedule or list]``
```

1ucian0 reviewed (commented):
LGTM. A very impressive first contribution.

since the `qubit_channel_mapping` deprecation was introduced by @nkanazawa1989 in https://github.com/Qiskit/qiskit/pull/8276, maybe better if he also have a look.

Similarly, https://github.com/Qiskit/qiskit/pull/5128 by @mtreinish deprecatd the qobj support in QasmSimulatorPy. Is the removal as inteneded?

Thanks!

santhosh-gnse committed:
Update qiskit/providers/basicaer/qasm_simulator.py

Formatted the docstring of .run() fnction as per the review suggestion

Co-authored-by: Luciano Bello <bel@zurich.ibm.com>

santhosh-gnse committed:
Update releasenotes/notes/remove-deprecated-code-in-0.22.0-139fa09ee0cc6df9.yaml

Added extra detail on the version of ``qiskit-terra`` from which ``qubit_channel_mapping`` was depreciated

Co-authored-by: Luciano Bello <bel@zurich.ibm.com>

santhosh-gnse committed:
Update releasenotes/notes/remove-deprecated-code-in-0.22.0-139fa09ee0cc6df9.yaml

Better formatted the release notes

Co-authored-by: Luciano Bello <bel@zurich.ibm.com>

santhosh-gnse commented:
> LGTM. A very impressive first contribution.
> 
> since the `qubit_channel_mapping` deprecation was introduced by @nkanazawa1989 in #8276, maybe better if he also have a look.
> 
> Similarly, #5128 by @mtreinish deprecatd the qobj support in QasmSimulatorPy. Is the removal as inteneded?
> 
> Thanks!

Thank you for your kind words; they mean a lot. That was great encouragement for me. Special thanks to @ElePT for their guidance—it made this contribution possible. Excited to contribute more and be part of the community!

santhosh-gnse committed:
Merge branch 'main' into remove-deprecate

santhosh-gnse committed:
Merge branch 'main' into remove-deprecate

mtreinish commented on qiskit/providers/basicaer/qasm_simulator.py:
```diff
@@ -387,18 +385,11 @@ def _validate_measure_sampling(self, experiment):
 
         Args:
-            qobj (Qobj): payload of the experiment
+            run_input (QuantumCircuit or Schedule or list): payload of the experiment
```
A schedule isn't supported on BasicAer, it's just a circuit or a list of circuits:

```suggestion
            run_input (QuantumCircuit or list): payload of the experiment
```

santhosh-gnse replied:
I wanted to express my gratitude for your approval! You're absolutely right; the lack of schedule support was an oversight on my part. I'll make sure to be more careful in my future PRs. Your feedback and attention to detail are highly valuable. Special thanks to @1ucian0 for updating the docstring.

mtreinish reviewed (commented):
This LGTM, thanks for taking care of this. I just caught one small mistake in the updated docstring for basicaer that I commented on inline.

I also just realized that `qiskit.providers.basicaer.UnitarySimulatorPy.run()` (https://github.com/Qiskit/qiskit/blob/main/qiskit/providers/basicaer/unitary_simulator.py#L206) never deprecated a Qobj input and will still support it while the qasm simulator and statevector simulator will not. I still think we should move forward with your PR, but it's something we'll want to clean up in a subsequent PR.

1ucian0 commented:
> I still think we should move forward with your PR, but it's something we'll want to clean up in a subsequent PR.

https://github.com/Qiskit/qiskit/issues/11046

1ucian0 committed:
Update qiskit/providers/basicaer/qasm_simulator.py

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>
